bemani/backend/iidx/cannonballers.py

```diff
@@ -1,15 +1,2184 @@
+                self.GAME_DP_DAN_RANK_3_DAN: self.DAN_RANK_3_DAN,
+                self.GAME_DP_DAN_RANK_4_DAN: self.DAN_RANK_4_DAN,
+                self.GAME_DP_DAN_RANK_5_DAN: self.DAN_RANK_5_DAN,
+                self.GAME_DP_DAN_RANK_6_DAN: self.DAN_RANK_6_DAN,
+                self.GAME_DP_DAN_RANK_7_DAN: self.DAN_RANK_7_DAN,
+                self.GAME_DP_DAN_RANK_8_DAN: self.DAN_RANK_8_DAN,
+                self.GAME_DP_DAN_RANK_9_DAN: self.DAN_RANK_9_DAN,
+                self.GAME_DP_DAN_RANK_10_DAN: self.DAN_RANK_10_DAN,
+                self.GAME_DP_DAN_RANK_CHUDEN: self.DAN_RANK_CHUDEN,
+                self.GAME_DP_DAN_RANK_KAIDEN: self.DAN_RANK_KAIDEN,
+            }[game_dan]
+        else:
+            raise Exception('Invalid cltype!')
+
+    def handle_IIDX25shop_getname_request(self, request: Node) -> Node:
+        machine = self.data.local.machine.get_machine(self.config['machine']['pcbid'])
+        if machine is not None:
+            machine_name = machine.name
+            close = machine.data.get_bool('close')
+            hour = machine.data.get_int('hour')
+            minute = machine.data.get_int('minute')
+        else:
+            machine_name = ''
+            close = False
+            hour = 0
+            minute = 0
+
+        root = Node.void('IIDX25shop')
+        root.set_attribute('opname', machine_name)
+        root.set_attribute('pid', '51')
+        root.set_attribute('cls_opt', '1' if close else '0')
+        root.set_attribute('hr', str(hour))
+        root.set_attribute('mi', str(minute))
+        return root
+
+    def handle_IIDX25shop_savename_request(self, request: Node) -> Node:
+        self.update_machine_name(request.attribute('opname'))
+
+        shop_close = intish(request.attribute('cls_opt')) or 0
+        minutes = intish(request.attribute('mnt')) or 0
+        hours = intish(request.attribute('hr')) or 0
+
+        self.update_machine_data({
+            'close': shop_close != 0,
+            'minutes': minutes,
+            'hours': hours,
+        })
+
+        return Node.void('IIDX25shop')
+
+    def handle_IIDX25shop_sentinfo_request(self, request: Node) -> Node:
+        return Node.void('IIDX25shop')
+
+    def handle_IIDX25shop_sendescapepackageinfo_request(self, request: Node) -> Node:
+        root = Node.void('IIDX25shop')
+        root.set_attribute('expire', str((Time.now() + 86400 * 365) * 1000))
+        return root
+
+    def handle_IIDX25shop_getconvention_request(self, request: Node) -> Node:
+        root = Node.void('IIDX25shop')
+        machine = self.data.local.machine.get_machine(self.config['machine']['pcbid'])
+        if machine.arcade is not None:
+            course = self.data.local.machine.get_settings(machine.arcade, self.game, self.music_version, 'shop_course')
+        else:
+            course = None
+
+        if course is None:
+            course = ValidatedDict()
+
+        root.set_attribute('music_0', str(course.get_int('music_0', 20032)))
```

Comment:
These are valid song IDs for songs found in CB, right? Just checking that they weren't copied without verifying.

Reply:
Checked just now and they are in the musicdb